### PR TITLE
[AOTAutograd] tweak min-cut partitioner to avoid saving softmax output

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #126348
* #126477

Right now the linear + cross entropy loss operation (usually to be the last part of a transformer model) does the following thing
1. run matmul to get softmax_input
2. load softmax_input to compute max per row.
3. load softmax_input to compute sum per row
4. load softmax_input, normalize it and save the result to softmax_output

Step 4 is inefficient since
a. in the fwd pass, only a small slice of the softmax_output tensor is need to compute NLLLoss. Materializing the whole tensor is an overkill
b. in the backward pass, we need the whole softmax_output, but it can be recompute from softmax_input

If we skip saving softmax_output, we would have perf wins since this is the largest tensor in the network. For llm.c, the size is batch_size * sequence_length * vocab_size * item_size ~= 32 * 1024 * 50257 * 2 ~= 3GB. Simply read/write such large tensor need ~2ms in A100. If we recompute softmax_output, we save 1 load for softmax_input and 1 store for softmax_output, which would result in ~4ms saving.

To avoid saving the softmax_output we need make sure the min cut partitioner decides to recompute it based on softmax_input and the max/sum tensor (which is small) computed in step 2 and 3. This is not happening currently since the min cut partitioner over-estimate the cost of recomputation.

The fix is suggested by @Chillee  to let `dist_from_bw` play a less important role.

### CODE REVIEW

**Review by Chillee — COMMENTED** [review]
Please add a property-based test in `test/inductor/test_perf.py`. I'd prefer that to the test currently in the PR.

Also, obviously, do a perf run.

**Comment by shunting314** [comment]
With the new heuristics, the backward runs slower and we end up with roughly neutral perf overall for llm.c. The reason is the kernel computing gradient of softmax input in the backward pass picks a sub-optimal triton config. #126477 fixes that, and now we have the 4ms saving as estimated in the summary for llm.c:

```
Latency: 194.20ms -> 190.15ms ,
Tokens/s 168.7K -> 172.3 K (2% improvement)
```

**Comment by shunting314** [comment]
> Also, obviously, do a perf run.

Perf run shows 5 seconds regress for TIMM [link](https://hud.pytorch.org/benchmark/compilers?startTime=Sat%2C%2011%20May%202024%2001%3A16%3A28%20GMT&stopTime=Sat%2C%2018%20May%202024%2001%3A16%3A28%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/146/head&lCommit=74f1a4a4c1fcf8944ca4c6237424eb03f2d55ac2&rBranch=main&rCommit=d61a81a9e76688ac8f338a6cfba932bf7779e5ce). I'll need debug where that comes from.

File: test/functorch/test_aotdispatch.py [hunk 1]
```diff
@@ -4892,6 +4898,71 @@ def test_autocast(self):
...
+        In transformer, it usually hurts perf if the partitioner saves softmax
+        output for the backward pass. We can skip writing the softmax output
+        but recompute it based on softmax input and max/sum statistics.
+        """
+        B = 32
+        T = 1024
+        C = 768
+        V = 50257
+
+        linear = nn.Linear(C, V, bias=False, dtype=torch.bfloat16).cuda()
```
Comment (etaf) [inline test/functorch/test_aotdispatch.py:4913]: Hi, may I suggest we replace `"cuda"` with the GPU_TYPE  in this case?  XPU also meet the  `has_triton` requirement.  You may also skip if not torch.cuda.is_available() for this case. Thanks.
